### PR TITLE
[fix,feat] Add loss for in-batch cross entropy and allow "target" to be specified on the fly

### PR DESCRIPTION
**Pull request by suzyahyah**
[Converted to draft. Metrics still requires targets]

Summary:
Previously, losses require "targets" in sample_list, and will silently return `{}` if "targets" are not present in `sample_list`, but this might not be the case for custom loss functions.

Self-supervised/metric-learning/retrieval based functions like in batch cross entropy (example provided) or alignment losses compute their targets on the fly within the batch, and leaving this flexible allows for different targets to be specified for different loss functions on the fly.

Differential Revision: D23269219

### CODE REVIEW

File: mmf/modules/losses.py [hunk 1]
```diff
@@ -559,3 +558,20 @@ def __init__(self, params=None):
...
+        if params is None:
+            params = {}
+        self.loss_fn = nn.CrossEntropyLoss(**params)
+
+    def forward(self, sample_list, model_output):
+        bs = len(sample_list['id'])
+        scores = model_output['scores']
+        assert scores.shape[0] == scores.shape[1]
+        # our targets are just along the diagonal
+        targets = torch.arange(0, bs, dtype=torch.long).to(device=scores.device)
```
Comment (vedanuj) [inline mmf/modules/losses.py:576]: This seems to depend only on the batch size. I think this can be done in the model forward pass as well and you can add it to the sample_list.
